File: binary-search/python/binarySearch.py
# Note: for repeating numbers, this algorithm will not
# always be correct. In order to make it so, one would
# need to include additional steps to ensure that the
# found value is lower than the one before it, and, if
# necessary, find the first repeating value (for example,
# iterating backwards until it is found)

def search(collection, target):
    if collection is None or not target:
        raise TypeError("Arguments cannot be of type None")
    if len(collection) == 0:
        return -1

    low = 0
    high = len(collection) - 1

    while low <= high:
        middle = (low + high) // 2
        guess = collection[middle]
        
        if (guess == target):
            return middle
        elif (collection[middle] > target):
            high = middle - 1

        else: 
            low = middle + 1

    return -1

File: java/src/algorithms/searching/BinarySearch.java
package algorithms.searching;

import static java.util.Objects.requireNonNull;

/**
 * Note: for repeating numbers, this algorithm will not always find a first
 * occurrence. In order to make it so, one would need to include additional
 * steps to ensure that the found value is lower than the one before it, and,
 * if necessary, find the first repeating value (for example, iterating
 * backwards until it is found)
 * */
public class BinarySearch {

    private BinarySearch() {
        throw new AssertionError("The class should not be instantiated");
    }

    public static <E extends Comparable<E>> int search(E[] collection, E target) {
        requireNonNull(collection, "Collection should not be null");
        requireNonNull(target, "Target object should not be null");

        int bottom = 0;
        int top = collection.length - 1;

        while (bottom <= top) {
            int mid = (top + bottom) / 2;
            if (collection[mid].compareTo(target) == 0) {
                return mid;
            } else if (collection[mid].compareTo(target) > 0) {
                top = mid -1;
            } else {
                bottom = mid + 1;
            }
        }

        return -1;
    }
}